From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Steve Anton <anxuiz.nx@gmail.com>
Date: Thu, 3 Mar 2016 00:09:38 -0600
Subject: [PATCH] Add PlayerInitialSpawnEvent

For modifying a player's initial spawn location as they join the server

This is a duplicate API from spigot, so use our duplicate subclass and
improve setPosition to use raw

diff --git a/src/main/java/net/minecraft/server/PlayerList.java b/src/main/java/net/minecraft/server/PlayerList.java
index 213e8512247e9873563ad606a7e3a2a3ca409718..2265fa3c04305e18653b4b37a0d1d36b261521d7 100644
--- a/src/main/java/net/minecraft/server/PlayerList.java
+++ b/src/main/java/net/minecraft/server/PlayerList.java
@@ -143,7 +143,7 @@ public abstract class PlayerList {
 
         // Spigot start - spawn location event
         Player bukkitPlayer = entityplayer.getBukkitEntity();
-        PlayerSpawnLocationEvent ev = new PlayerSpawnLocationEvent(bukkitPlayer, bukkitPlayer.getLocation());
+        PlayerSpawnLocationEvent ev = new com.destroystokyo.paper.event.player.PlayerInitialSpawnEvent(bukkitPlayer, bukkitPlayer.getLocation()); // Paper use our duplicate event
         Bukkit.getPluginManager().callEvent(ev);
 
         Location loc = ev.getSpawnLocation();
@@ -151,7 +151,7 @@ public abstract class PlayerList {
 
         entityplayer.spawnIn(worldserver1);
         entityplayer.playerInteractManager.a((WorldServer) entityplayer.world);
-        entityplayer.setPosition(loc.getX(), loc.getY(), loc.getZ());
+        entityplayer.setPositionRaw(loc.getX(), loc.getY(), loc.getZ()); // Paper - set raw so we aren't fully joined to the world (not added to chunk or world)
         entityplayer.setYawPitch(loc.getYaw(), loc.getPitch());
         // Spigot end
 
